common/txmgr/types/storage_service.go

```diff
@@ -0,0 +1,102 @@
+	// FindTransactionsConfirmedInBlockRange(highBlockNumber, lowBlockNumber int64, chainID CHAINID, opts ...any) (etxs []*TX, err error)
+	// GetEthTxInProgress(fromAddress ADDR, opts ...any) (etx *TX, err error)
+	// GetInProgressEthTxAttempts(ctx context.Context, address ADDR, chainID CHAINID, opts ...any) (attempts []TXATTEMPT, err error)
+	// HasInProgressTransaction(account ADDR, chainID CHAINID, opts ...any) (exists bool, err error)
+	// // InsertEthReceipt only used in tests. Use SaveFetchedReceipts instead
+	InsertEthReceipt(receipt *Receipt[R, HASH]) error
+	// InsertEthTx(etx *TX, opts ...any) error
+	// InsertEthTxAttempt(attempt *TXATTEMPT, opts ...any) error
+	// LoadEthTxAttempts(etx *TX, opts ...any) error
+	// LoadEthTxesAttempts(etxs []*TX, opts ...any) error
+	// MarkAllConfirmedMissingReceipt(chainID CHAINID, opts ...any) (err error)
+	// MarkOldTxesMissingReceiptAsErrored(blockNum int64, finalityDepth uint32, chainID CHAINID, opts ...any) error
+	// PreloadEthTxes(attempts []TXATTEMPT, opts ...any) error
+	// SaveConfirmedMissingReceiptAttempt(ctx context.Context, timeout time.Duration, attempt *TXATTEMPT, broadcastAt time.Time, opts ...any) error
+	SaveFetchedReceipts(receipts []R, chainID CHAINID) (err error)
+	// SaveInProgressAttempt(attempt *TXATTEMPT, opts ...any) error
+	// SaveInsufficientEthAttempt(timeout time.Duration, attempt *TXATTEMPT, broadcastAt time.Time, opts ...any) error
+	// SaveReplacementInProgressAttempt(oldAttempt TXATTEMPT, replacementAttempt *TXATTEMPT, opts ...any) error
+	// SaveSentAttempt(timeout time.Duration, attempt *TXATTEMPT, broadcastAt time.Time, opts ...any) error
+	// SetBroadcastBeforeBlockNum(blockNum int64, chainID CHAINID, opts ...any) error
+	// UpdateBroadcastAts(now time.Time, etxIDs []TXID, opts ...any) error
+	// UpdateEthKeyNextNonce(newNextNonce, currentNextNonce TXMETA, address ADDR, chainID CHAINID, opts ...any) error
+	// UpdateEthTxAttemptInProgressToBroadcast(etx *TX, attempt TXATTEMPT, NewAttemptState TxAttemptState, incrNextNonceCallback CallbackFunc, opts ...any) error
+	// UpdateEthTxsUnconfirmed(ids []int64, opts ...any) error
+	// UpdateEthTxUnstartedToInProgress(etx *TX, attempt *TXATTEMPT, opts ...any) error
+	// UpdateEthTxFatalError(etx *TX, opts ...any) error
+	// UpdateEthTxForRebroadcast(etx TX, etxAttempt TXATTEMPT, opts ...any) error
+	// Close()
+}
+
+// TxStrategy controls how txes are queued and sent
+//
+//go:generate mockery --quiet --name TxStrategy --output ./mocks/ --case=underscore --structname TxStrategy --filename tx_strategy.go
+type TxStrategy interface {
+	// Subject will be saved txes.subject if not null
+	Subject() uuid.NullUUID
+	// PruneQueue is called after tx insertion
+	// It accepts the service responsible for deleting
+	// unstarted txs and deletion options
+	PruneQueue(pruneService PruneService, qopt pg.QOpt) (n int64, err error)
+}
+
+type PruneService interface {
```

Comment:
Or since this is just one method, you could make a `type PruneUnstartedTxQueueFn func (queueSize uint32, subject uuid.UUID, qopts ...pg.QOpt) (n int64, err error)` and just pass the method reference `.PruneUnstartedTxQueue`

Reply:
I prefer the interface approach but good point about the naming

Reply:
+1 to renaming away from PruneService. That name currently indicates it is a separate service.

Since in future every chain family would implement it differently, I prefer an interface over func approach. But no strong preferences.

Reply:
The func approach doesn't limit anything if there will only be one method. You can still pass method references as the func type just the same. Not really significant either way though :shrug: